File: backend/src/main/java/com/org/playboard/dto/stats/LeaderboardEntryDto.java
package com.org.playboard.dto.stats;

import java.math.BigDecimal;
import java.util.List;
import java.util.UUID;

/**
 * One row of the leaderboard.
 *
 * <p>{@code rating} is the cumulative team-v2 skill score mapped to 0-100 with one decimal.
 * {@code provisional} players have played fewer than
 * the group's {@link LeaderboardResponse#minGamesToRank()} and are listed after the ranked
 * ones; they still carry a continuing {@code rank} rather than a sentinel, so older clients
 * that don't know about the flag still render a sanely numbered list.
 *
 * <p>{@code recentForm} is the player's last 10 results within the standings window, in
 * <b>chronological</b> order (oldest first, newest last) — deliberately the opposite of
 * {@link PlayerStatsDto#recentMatches()}, which is newest-first. Chronological order is the
 * render order for a left-to-right form-dots row, so no client needs to reverse it.
 */
public record LeaderboardEntryDto(
        int rank,
        UUID userId,
        String displayName,
        String photoUrl,
        String avatarId,
        String avatarColor,
        int gamesPlayed,
        int wins,
        int losses,
        int pointsFor,
        int pointsAgainst,
        BigDecimal winRate,
        int currentStreak,
        int bestStreak,
        BigDecimal rating,
        boolean provisional,
        List<Boolean> recentForm,
        String algorithmVersion,
        String ratingPeriod,
        int uniquePartners,
        BigDecimal maxPartnerShare,
        String provisionalReason,
        boolean limitedPartnerVariety) {

    public LeaderboardEntryDto(
            int rank, UUID userId, String displayName, String photoUrl, String avatarId,
            String avatarColor, int gamesPlayed, int wins, int losses, int pointsFor,
            int pointsAgainst, BigDecimal winRate, int currentStreak, int bestStreak,
            BigDecimal rating, boolean provisional, List<Boolean> recentForm) {
        this(rank, userId, displayName, photoUrl, avatarId, avatarColor, gamesPlayed, wins,
                losses, pointsFor, pointsAgainst, winRate, currentStreak, bestStreak, rating,
                provisional, recentForm, "wilson-v1", "all-time", 0, BigDecimal.ZERO, null, false);
    }

    /** Compatibility constructor for callers compiled against the pre-warning payload. */
    public LeaderboardEntryDto(int rank, UUID userId, String displayName, String photoUrl, String avatarId,
            String avatarColor, int gamesPlayed, int wins, int losses, int pointsFor, int pointsAgainst,
            BigDecimal winRate, int currentStreak, int bestStreak, BigDecimal rating, boolean provisional,
            List<Boolean> recentForm, String algorithmVersion, String ratingPeriod, int uniquePartners,
            BigDecimal maxPartnerShare, String provisionalReason) {
        this(rank, userId, displayName, photoUrl, avatarId, avatarColor, gamesPlayed, wins, losses, pointsFor,
                pointsAgainst, winRate, currentStreak, bestStreak, rating, provisional, recentForm,
                algorithmVersion, ratingPeriod, uniquePartners, maxPartnerShare, provisionalReason, false);
    }

    /**
     * Copy with a different rank. Lives on the record so the 17-field constructor is
     * splatted in exactly one place — positional construction at several call sites was
     * a standing hazard every time a field was added.
     */
    public LeaderboardEntryDto withRank(int newRank) {
        return new LeaderboardEntryDto(
                newRank, userId, displayName, photoUrl, avatarId, avatarColor,
                gamesPlayed, wins, losses, pointsFor, pointsAgainst,
                winRate, currentStreak, bestStreak, rating, provisional, recentForm,
                algorithmVersion, ratingPeriod, uniquePartners, maxPartnerShare, provisionalReason, limitedPartnerVariety);
    }

    /** Copy marked provisional (or not). */
    public LeaderboardEntryDto withProvisional(boolean value) {
        return new LeaderboardEntryDto(
                rank, userId, displayName, photoUrl, avatarId, avatarColor,
                gamesPlayed, wins, losses, pointsFor, pointsAgainst,
                winRate, currentStreak, bestStreak, rating, value, recentForm,
                algorithmVersion, ratingPeriod, uniquePartners, maxPartnerShare, provisionalReason, limitedPartnerVariety);
    }

    /** Points difference — the first tiebreak between equal ratings, and shown on the row. */
    public int pointsDiff() {
        return pointsFor - pointsAgainst;
    }
}

File: android/app/src/main/java/com/org/playboard/data/remote/dto/LeaderboardEntryDto.kt
package com.org.playboard.data.remote.dto

import kotlinx.serialization.Serializable

/**
 * One row of `GET /groups/{groupId}/leaderboard` — server-sorted by rating desc,
 * then points difference desc, then wins desc
 * (docs/backend/api-contracts.md § Leaderboard).
 */
@Serializable
data class LeaderboardEntryDto(
    val rank: Int,
    val userId: String,
    val displayName: String,
    val photoUrl: String?,
    val avatarId: String? = null,
    val avatarColor: String,
    val gamesPlayed: Int,
    val wins: Int,
    val losses: Int,
    val pointsFor: Int,
    // Defaulted so a pre-points-difference backend's JSON still deserializes during
    // rollout (diff then reads as pointsFor, until the deploy lands).
    val pointsAgainst: Int = 0,
    val winRate: Double,
    // Signed: positive = current win streak, negative = current loss streak.
    // Defaulted so a pre-streak backend's JSON still deserializes during rollout.
    val currentStreak: Int = 0,
    val bestStreak: Int = 0,
    // Wilson score lower bound on the win rate, 0-100. Nullable rather than defaulted to
    // 0.0 — unlike the fields above, 0.0 is a *legitimate* rating for a winless player, so
    // a default would make "backend predates ratings" indistinguishable from "has never
    // won". The UI falls back to showing win% while this is null.
    val rating: Double? = null,
    // Below the group's minGamesToRank: listed, but not ranked.
    val provisional: Boolean = false,
)
